### PR TITLE
fixing edge cases for slicing + unit tests.

### PR DESCRIPTION
**Pull request by weinbe58**

### CODE REVIEW

File: lib/BloqadeWaveforms/test/waveform.jl [hunk 1]
```diff
@@ -192,7 +200,9 @@ end
     pwl = piecewise_linear(;clocks=Float64[0,1],values=Float64[1,4])
     @test pwl[0.0..0.5] ≈ piecewise_linear(clocks=Float64[0,0.5],values=Float64[1,2.5])
     @test pwl[0.5..1.0] ≈ piecewise_linear(clocks=Float64[0,0.5],values=Float64[2.5,4])
+    @test pwl[0.0..0.0] ≈ piecewise_linear(clocks=Float64[0,0.0],values=Float64[1,1])
```
Comment (johnzl-777) [inline lib/BloqadeWaveforms/test/waveform.jl:203]: This seems to cause `Interpolations.jl` to issue warnings:

```
┌ Warning: Duplicated knots were deduplicated. Use Interpolations.deduplicate_knots!(knots) explicitly to avoid this warning.
│   k1 =
│    2-element Vector{Float64}:
│     0.0
│     0.0
└ @ Interpolations ~/.julia/packages/Interpolations/y4lLj/src/gridded/gridded.jl:77
```

File: lib/BloqadeWaveforms/test/waveform.jl [hunk 2]
```diff
@@ -167,6 +167,12 @@ end
     end
 end
 
+begin
```
Comment (johnzl-777) [inline lib/BloqadeWaveforms/test/waveform.jl:170]:
```suggestion
```

File: lib/BloqadeWaveforms/test/waveform.jl [hunk 3]
```diff
@@ -167,6 +167,12 @@ end
...
+
+
+end
```
Comment (johnzl-777) [inline lib/BloqadeWaveforms/test/waveform.jl:173]:
```suggestion
```
Would remove all the associated whitespace as well between `begin` and `end` here

File: lib/BloqadeWaveforms/test/waveform.jl (hunk 1 above)
Comment (johnzl-777) [inline lib/BloqadeWaveforms/test/waveform.jl:203]: Both parts of the approximate equality (`pwl[0.0..0.0]` and `piecewise_linear(...)`) cause `Interpolations.jl` to issue the following warning: 
```
┌ Warning: Duplicated knots were deduplicated. Use Interpolations.deduplicate_knots!(knots) explicitly to avoid this warning.
│   k1 =
│    2-element Vector{Float64}:
│     0.0
│     0.0
└ @ Interpolations ~/.julia/packages/Interpolations/y4lLj/src/gridded/gridded.jl:77
```

**Review by johnzl-777 — CHANGES_REQUESTED** [review]
Good that all test cases pass but I'll dig into why `Interpolations.jl` doesn't seem to be happy.

**Comment by johnzl-777** [comment]
After some digging around it seems the call to `LinearInterpolation` in `piecewise_linear` does not like the fact that the clocks and values can be the same (e.g. clocks = [0.0, 0.0], values = [1.0, 1.0]) in the case of a waveform slice with the same starting and ending interval values.

Digging deeper, the only time the `interp` field in the PiecewiseLinear struct ever gets used is when a user calls `(f::PiecewiseLinear)(t::Real) = f.interp(t)`. Perhaps it might be beneficial to introduce a type union to what `interp` can be and define something that can just handle the case where a piecewise linear only has one clock value twice (although I'm not sure if that even makes sense within the greater context of waveform manipulation).

**Comment by weinbe58** [comment]
I think some of the warnings are unavoidable but there are some cases which we can remove the extra knots

**Comment by johnzl-777** [comment]
@weinbe58 I'll give the thumbs up on this one although I think it's also worth either putting a notice in the docs that the warning is normal and/or finding a way to suppress this specific warning in the future.

**Review by johnzl-777 — APPROVED** [review]
